arbor/morph/morphology.cpp

```diff
@@ -0,0 +1,208 @@
+#include <cmath>
+#include <iostream>
+#include <unordered_map>
+#include <utility>
+
+#include <arbor/math.hpp>
+#include <arbor/morph/error.hpp>
+#include <arbor/morph/morphology.hpp>
+#include <arbor/morph/sample_tree.hpp>
+#include <arbor/morph/primitives.hpp>
+
+#include "algorithms.hpp"
+#include "util/span.hpp"
+#include "util/strprintf.hpp"
+
+namespace arb {
+
+namespace impl{
+
+std::vector<mbranch> branches_from_parent_index(const std::vector<size_t>& parents, const std::vector<point_prop>& props, bool spherical_root) {
+    using util::make_span;
+
+    const char* errstr_single_sample_root =
+        "A morphology with only one sample must have a spherical root";
+    const char* errstr_incomplete_cable =
+        "A branch must contain at least two samples";
+
+    auto npos = mbranch::npos;
+
+    if (!parents.size()) return {};
+
+    auto nsamp = parents.size();
+
+    // Enforce that a morphology with one sample has a spherical root.
+    if (!spherical_root && nsamp==1u) {
+        throw morphology_error(errstr_single_sample_root);
+    }
+
+    std::vector<int> bids(nsamp);
+    int nbranches = spherical_root? 1: 0;
+    for (auto i: make_span(1, nsamp)) {
+        size_t p = parents[i];
+        bool first = is_root(props[p]) || is_fork(props[p]);
+        bids[i] = first? nbranches++: bids[p];
+    }
+
+    std::vector<mbranch> branches(nbranches);
+    for (auto i: make_span(nsamp)) {
+        auto p = parents[i];
+        auto& branch = branches[bids[i]];
+
+        // Determine the id of the parent branch if this is the first sample in
+        // the branch, and include the fork/root point at the end of the parent
+        // branch where applicable.
+        // This is icky, but of all the solutions, this is the cleanest (that I
+        // could find).
+        if (!branch.size()) {
+            // A branch has null root if either:
+            //      ∃ a spherical root and branch id is 0
+            //      ∄ a spherical root and parent id is 0
+            auto null_root = spherical_root? !i: !p;
+            branch.parent_id = null_root? npos: bids[parents[i]];
+
+            // Add the first sample to a branch that is attached to
+            // non-spherical root if the branch is not branch 0.
+            if ((null_root && i) || is_fork(props[p])) {
+                branch.index.push_back(p);
+            }
+        }
+        branch.index.push_back(i);
+    }
+
+    // Enforce that all cable branches that are potentially connected to a spherical
+    // root contain at least two samples.
+    if (spherical_root) {
+        for (auto i: make_span(1, nbranches)) { // skip the root.
+            if (branches[i].size()<2u) {
+                throw morphology_error(errstr_incomplete_cable);
+            }
+        }
+    }
+
+    return branches;
+}
+
+// Returns false if one of the root's children has the same tag as the root.
+bool has_single_root_tag(const sample_tree& st) {
+    if (!st.size()) return false;
```

Comment:
I didn't use `unique`, because the tag on the root doesn't have to be unique.
I am not too fussed: it is a helper in `impl` namespace that is contained inside the TU.

Reply:
I should be clearer: the tag on the root sample does not need to be unique over the whole morphology: it only has to be unique for the root and the roots direct children.
I renamed it `root_sample_has_same_tag_as_child`